### PR TITLE
fixed invalid share

### PR DESCRIPTION
**Pull request by hackmod**
this is a quick hack to fix invalid share.
manually partial revert commit "Simplify logic for new_work checks" a4e74a0518a8

***only CUDA tested***

thanks for your great work!!

### CODE REVIEW

File: libethash-cl/CLMiner.cpp [hunk 1]
```diff
@@ -758,22 +757,19 @@ void CLMiner::ethash_search()
...
+    // process batches until we get new work.
+    bool done = false;
 
     while (!done)
     {
         // Exit next time around if there's new work awaiting
-        if (!done)
-            done = m_new_work.load(memory_order_relaxed);
+        bool t = true;
+        done = (m_new_work.compare_exchange_strong(t, false) || paused() || shouldStop());
```
Comment (StefanOberhumer) [inline libethash-cl/CLMiner.cpp]: I don't think check for paused() is needed in the inner xxx_search() loop.
If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

File: libethash-cl/CLMiner.cpp [hunk 2]
```diff
@@ -892,22 +887,19 @@ void CLMiner::progpow_search()
...
```
Comment (StefanOberhumer) [inline libethash-cl/CLMiner.cpp]: I don't think check for paused() is needed in the inner xxx_search() loop.
If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

File: libethash-cpu/CPUMiner.cpp [hunk 3]
```diff
@@ -240,7 +241,8 @@ void dev::eth::CPUMiner::progpow_search()
     while (true)
     {
         // Exit next time around if there's new work awaiting
-        if (m_new_work.load(memory_order_relaxed))
+        bool t = true;
+        if (m_new_work.compare_exchange_strong(t, false) || paused() || shouldStop())
```
Comment (StefanOberhumer) [inline libethash-cpu/CPUMiner.cpp]: I don't think check for paused() is needed in the inner xxx_search() loop.
If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

File: libethash-cuda/CUDAMiner.cpp [hunk 4]
```diff
@@ -402,28 +402,33 @@ void CUDAMiner::ethash_search()
...
+
     while (!done)
     {
+        // Exit next time around if there's new work awaiting
+        bool t = true;
+        done = m_new_work.compare_exchange_strong(t, false);
+
+        // Check on every batch if we need to suspend mining
+        if (!done)
+            done = paused();
```
Comment (StefanOberhumer) [inline libethash-cuda/CUDAMiner.cpp]: I don't think check for paused() is needed in the inner xxx_search() loop.
If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

File: libethash-cuda/CUDAMiner.cpp [hunk 5]
```diff
@@ -498,34 +511,38 @@ void CUDAMiner::progpow_search()
...
+    bool done = false;
     while (!done)
     {
+        // Exit next time around if there's new work awaiting
+        bool t = true;
+        done = m_new_work.compare_exchange_strong(t, false);
+
+        // Check on every batch if we need to suspend mining
+        if (!done)
+            done = paused();
```
Comment (StefanOberhumer) [inline libethash-cuda/CUDAMiner.cpp]: I don't think check for paused() is needed in the inner xxx_search() loop.
If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

**Comment by AndreaLanfranchi** [comment]
@hackmod thank you for jumping on this problem.
I'm quite astonished the mods you have reverted may be the causes of invalid shares: nevertheless I will give it a try.

You have tested it ? No invalids at all ?

**Comment by hackmod** [comment]
> @StefanOberhumer wrote:
> I don't think check for paused() is needed in the inner xxx_search() loop.
> If pause gets set (`Miner::pause(MinerPauseEnum what)`) `kick_miner()` gets called which sets `m_new_work`.

I do not try to modify original commit a4e74a0518a8965cffa29e3, this fix is based on the `git bisect` result and partially reverted manually.

> @hackmod thank you for jumping on this problem.
> I'm quite astonished the mods you have reverted may be the causes of invalid shares: nevertheless I will give it a try.
> 
> You have tested it ? No invalids at all ?

yes. for a private chain with low difficulty, I can observe invalid shares almost immediately,
but with this quick hack, no invalid shares observed for hours.

**Comment by AndreaLanfranchi** [comment]
> yes. for a private chain with low difficulty, I can observe invalid shares almost immediately,
but with this quick hack, no invalid shares observed for hours.

Thank you for feed back.
This puzzles me as none of the modifications reverted has something to do with the number of streams.
(with 1 stream only everything is working while from 2+ streams I begin to detect invalids).

Scratching my head...

**Comment by AndreaLanfranchi** [comment]
@hackmod unfortunately I've tested it and must say your reverts do not solve the problem.
It's proven concurrent streams **do produce** invalid results even on original ifdefelse's implementation.

Refer to this and following
https://github.com/ifdefelse/ProgPOW/issues/22#issuecomment-451661940

Probably, reverting some updates which are meant to speed up kernel launches, you have reintroduced (due to several checks against locking functions) small delays in kernel launches which reduce concurrency. But that's a workaround ... not the solution.

**Comment by hackmod** [comment]
> @hackmod unfortunately I've tested it and must say your reverts do not solve the problem.
> It's proven concurrent streams **do produce** invalid results even on original ifdefelse's implementation.
> 
> Refer to this and following
> [ifdefelse/ProgPOW#22 (comment)](https://github.com/ifdefelse/ProgPOW/issues/22#issuecomment-451661940)
> 
> Probably, reverting some updates which are meant to speed up kernel launches, you have reintroduced (due to several checks against locking functions) small delays in kernel launches which reduce concurrency. But that's a workaround ... not the solution.

I guess, it could be regarded as a CUDA's bug or something and..
when ideal solution is not working, workaround or some hack preferred temporary in my opionion.

**Pull request closed without merging**